### ISSUE TITLE
Macro place is not taken into account in variable name resolution

### ISSUE THREAD

**Issue by ChayimFriedman2** [rust-lang/rust-analyzer#18262]
In the following code:
```rust
fn foo() {
    let v: i32 = 0;
    macro_rules! m {
        () => {
            v
        };
    }
    let v: String = String::new();
    bar(m!());
}

fn bar(_: i32) {}
```
r-a reports an (experimental) diagnostics, because it infers `m!()` to `String`, because it resolves `v` in the macro to the second `v`:
```rust
expected i32, found String (rust-analyzer(E0308))
```

I'm going to work on this (because without this, my hygiene work generates a huge number of false positives), but it's nice to have an issue to close. So:
@rustbot claim

I expect this quite hard to do since we currently expand the macros during HIR lowering and then when we resolve them we have no idea that the second `v` came from a macro, but we'll see (I also haven't checked yet what rustc does).
